File: trunk/theLink/example/java/Filter1.java
/**
 *  \file       theLink/example/java/Filter1.java
 *  \brief      \$Id$
 *  
 *  (C) 2009 - NHI - #1 - Project - Group
 *  
 *  \version    \$Rev$
 *  \author     EMail: aotto1968 at users.berlios.de
 *  \attention  this software has GPL permissions to copy
 *              please contact AUTHORS for additional information
 */
package example;
import java.util.ArrayList;
import javamsgque.*;

class Filter1 extends MqS implements IFactory {
  private ArrayList<ArrayList<String>> data = new ArrayList<ArrayList<String>>();

  public MqS Factory() {
    return new Filter1();
  }

  public static class FTR implements IService {
    public void Service(MqS ctx) throws MqSException {
      ArrayList<String> d = new ArrayList<String>();
      while (ctx.ReadGetNumItems() != 0) {
	d.add("<" + ctx.ReadC() + ">");
      }
      ((Filter1)ctx).data.add(d);
      ctx.SendRETURN();
    }
  }

  public static class EOF implements IService {
    public void Service (MqS ctx) throws MqSException {
      MqS ftr = ctx.ConfigGetFilter();
      for (ArrayList<String> d: ((Filter1)ctx).data) {
	ftr.SendSTART();
	for (String s: d) {
	  ftr.SendC(s);
	}
	ftr.SendEND_AND_WAIT("+FTR");
      }
      ftr.SendSTART();
      ftr.SendEND_AND_WAIT("+EOF");
      ctx.SendRETURN();
    }
  }

  public static void main(String[] argv) {
    Filter1 srv = new Filter1();
    try {
      srv.ConfigSetName("filter");
      srv.ConfigSetIsServer(true);
      srv.LinkCreate(argv);
      srv.ServiceCreate("+FTR", new Filter1.FTR()); 
      srv.ServiceCreate("+EOF", new Filter1.EOF()); 
      srv.ProcessEvent(MqS.WAIT.FOREVER);
    } catch (Throwable e) {
      srv.ErrorSet(e);
    } finally {
      srv.Exit();
    }
  }
}

File: trunk/theLink/example/go/Filter1.go
/**
 *  \file       theLink/example/go/Filter1.go
 *  \brief      \$Id$
 *  
 *  (C) 2010 - NHI - #1 - Project - Group
 *  
 *  \version    \$Rev$
 *  \author     EMail: aotto1968 at users.berlios.de
 *  \attention  this software has GPL permissions to copy
 *              please contact AUTHORS for additional information
 */

package main

import (
  . "gomsgque"
    "os"
    //"fmt"
)

type Filter1 struct {
  *MqS
  data [][]string
}

func NewFilter1(tmpl *MqS) *MqS {
  return NewMqS(tmpl, new(Filter1))
}

func (this *Filter1) ServerSetup() {
  this.ServiceCreate("+FTR", (*FTR)(this))
  this.ServiceCreate("+EOF", (*EOF)(this))
}

type FTR Filter1
  func (this *FTR) Call() {
    var d []string
    for this.ReadItemExists() {
      d = append(d, "<" + this.ReadC() + ">")
    }
    this.data = append(this.data, d)
    this.SendRETURN()
  }

type EOF Filter1
  func (this *EOF) Call() {
    ftr := this.ServiceGetFilter()
    for _,d := range this.data {
      ftr.SendSTART()
      for _,s := range d {
	ftr.SendC(s)
      }
      ftr.SendEND_AND_WAIT2("+FTR")
    }
    ftr.SendSTART()
    ftr.SendEND_AND_WAIT2("+EOF")
    this.SendRETURN()
  }

func main() {
  srv := FactoryNew("Filter1", NewFilter1)
  defer func() {
    if x := recover(); x != nil {
      srv.ErrorSet(x)
    }
    srv.Exit()
  }()
  srv.LinkCreate(os.Args...)
  srv.ProcessEvent(WAIT_FOREVER)
}

